File: api/pkg/utils/liveness/liveness.go
// Package liveness computes whether a season/collection or a single item is on
// sale RIGHT NOW. It is the single source of that truth — status is COMPUTED on
// read from the date window + a manual status override. There is deliberately no
// cron and no persisted "live" boolean; nothing has to run for a season to open
// or close on time. Mirrored on the TS side at web/lib/store/liveness.ts.
package liveness

import "time"

// Collection statuses. draft/ended force OFF, live forces ON, scheduled derives
// from the window.
const (
	StatusDraft     = "draft"
	StatusScheduled = "scheduled"
	StatusLive      = "live"
	StatusEnded     = "ended"
)

// IsLiveNow reports whether a collection with the given status + window is
// currently on sale.
//
//   - draft / ended → false (manual override: force off)
//   - live          → true  (manual override: force on, ignore the window)
//   - anything else (scheduled) → true iff now is inside [saleStart, saleEnd],
//     where a nil bound means "open on that side"
func IsLiveNow(status string, saleStart, saleEnd *time.Time, now time.Time) bool {
	switch status {
	case StatusDraft, StatusEnded:
		return false
	case StatusLive:
		return true
	}
	return WindowContains(saleStart, saleEnd, now)
}

// ItemLiveNow reports whether a single item is sellable: it must be active and,
// if it carries its own window override, now must be inside it. Items have no
// status override — only active + optional window.
func ItemLiveNow(active bool, saleStart, saleEnd *time.Time, now time.Time) bool {
	return active && WindowContains(saleStart, saleEnd, now)
}

// WindowContains reports whether now falls in the half-open-ish window
// [saleStart, saleEnd] (inclusive both ends). A nil bound is open on that side.
func WindowContains(saleStart, saleEnd *time.Time, now time.Time) bool {
	startOk := saleStart == nil || !saleStart.After(now) // saleStart <= now
	endOk := saleEnd == nil || !saleEnd.Before(now)      // saleEnd   >= now
	return startOk && endOk
}

// StatusLabel is the label to SHOW in the admin UI (draft | scheduled | live |
// ended). It reflects both the override and the window so the operator sees the
// effective state, not just the stored string.
func StatusLabel(status string, saleStart, saleEnd *time.Time, now time.Time) string {
	switch status {
	case StatusDraft:
		return StatusDraft
	case StatusEnded:
		return StatusEnded
	case StatusLive:
		return StatusLive
	}
	if saleStart != nil && saleStart.After(now) {
		return StatusScheduled
	}
	if saleEnd != nil && saleEnd.Before(now) {
		return StatusEnded
	}
	return StatusLive
}

File: web/lib/store/liveness.ts
/**
 * Client mirror of the Go liveness util (api/pkg/utils/liveness). "On sale now"
 * is COMPUTED on read from the date window + a status override — there is no cron
 * and no persisted live flag. The admin list already gets status_label/is_live
 * from the API; this is used for live PREVIEW as the operator edits dates in the
 * editor, and by the timeline. Keep it in lockstep with the Go version.
 */

export type SeasonStatus = 'draft' | 'scheduled' | 'live' | 'ended';

/** now inside [start, end] (inclusive); a null/empty bound is open on that side. */
export function windowContains(start?: string | null, end?: string | null, now = Date.now()): boolean {
  const startOk = !start || new Date(start).getTime() <= now;
  const endOk = !end || new Date(end).getTime() >= now;
  return startOk && endOk;
}

/** draft/ended → off; live → on (override, ignore window); else window-derived. */
export function isLiveNow(status: string, start?: string | null, end?: string | null, now = Date.now()): boolean {
  if (status === 'draft' || status === 'ended') return false;
  if (status === 'live') return true;
  return windowContains(start, end, now);
}

/** Effective label to display (draft | scheduled | live | ended). */
export function statusLabel(status: string, start?: string | null, end?: string | null, now = Date.now()): SeasonStatus {
  if (status === 'draft') return 'draft';
  if (status === 'ended') return 'ended';
  if (status === 'live') return 'live';
  if (start && new Date(start).getTime() > now) return 'scheduled';
  if (end && new Date(end).getTime() < now) return 'ended';
  return 'live';
}

/** Thai label + brand color per status, shared by badges and the timeline bars. */
export const STATUS_META: Record<SeasonStatus, { label: string; className: string; bar: string }> = {
  live: { label: 'ขายอยู่', className: 'bg-accent-soft text-accent-ink', bar: 'bg-accent' },
  scheduled: { label: 'ตั้งเวลาไว้', className: 'bg-[#fde4d3] text-accent2', bar: 'bg-accent2' },
  ended: { label: 'จบแล้ว', className: 'bg-line text-soft', bar: 'bg-muted' },
  draft: { label: 'ร่าง', className: 'bg-line/60 text-muted', bar: 'bg-muted/60' },
};
